innerprod/fullysec/damgard.go

```diff
@@ -241,6 +262,17 @@ func (d *Damgard) Decrypt(cipher data.Vector, key *DamgardDerivedKey, y data.Vec
 	if err != nil {
 		return nil, err
 	}
-	return calc.WithBound(bound).BabyStepGiantStep(r, d.Params.g)
 
+	// TODO: this takes more time if the input is negative
+	res, err := calc.WithBound(bound).BabyStepGiantStep(r, d.Params.g)
```

Comment:
There are two things we could improve here:
1. See if there's a way to determine if the input is negative other than running the baby step-giant step algorithm once.
2. In the current implementation it works, but a non-nil error return value can signify other errors than "failed to find discrete logarithm within bound". This could pose a problem if we change the discrete log algorithm.

Reply:
1. I implemented a goroutine version for baby step-giant step.
2. For now I left this solution, since we probably do not plan to change much the baby step-giant step algorithm.